### PR TITLE
Add backend validations for CSV links

### LINKED ISSUE #648: do not create multiple tasks if csv not used
- if we attach a csv to the project it will create as many tasks as rows, that's ok if the csv columns are referenced but not if they arent

### PR DESCRIPTION
**Pull request by shirishgoyal**
Extends #650 
Closes #648

### CODE REVIEW

**Comment by dmorina** [comment]
ImportError: No module named django_orm
travis check fails
could you have a look, I think it's because of google-api-client version but you should check to be sure

**Comment by shirishgoyal** [comment]
Resolved django_orm issue. It was related to package version.

**Comment by dmorina** [comment]
lgtm

**Comment by dmorina** [comment]
buggy: 
screen shot 2016-03-04 at 10 49 31 am
clicking on publish doesnt fire the Dialog Open sometimes, and then if it does the default Task field is empty and publish doesnt work from there 

screen shot 2016-03-04 at 10 49 40 am


could you have a look?

File: crowdsourcing/serializers/project.py [hunk 1]
```diff
@@ -121,13 +124,44 @@ def get_comments(self, obj):
...
+            raise ValidationError('At least one template item is required')
+
+        has_input_item = False
+        for item in items.all():
+            if item.role == "input" or item.type == 'iframe':
+                has_input_item = True
+                break
+
+        if not has_input_item:
+            raise ValidationError('At least one input template item is required')
```
Comment (dmorina) [inline crowdsourcing/serializers/project.py]: validations should be serializer validators, check out DRF

File: crowdsourcing/serializers/project.py [hunk 2]
```diff
@@ -146,26 +149,52 @@ def get_comments(obj):
...
         # self.create_revision(None, *args, **kwargs)
         num_rows = self.validated_data.get('num_rows', 0)
+
         if self.instance.status != status and status == models.Project.STATUS_PUBLISHED:
             if self.instance.template.items.count() == 0:
                 raise ValidationError('At least one template item is required')
-            if self.instance.batch_files.count() == 0:
+
+            if self.instance.batch_files.count() == 0 or not self.has_csv_linkage(
+                    self.instance.templates.all()[0].template_items):
```
Comment (dmorina) [inline crowdsourcing/serializers/project.py]: self.instance.template.items

File: crowdsourcing/validators/project.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    def set_context(self, serializer):
+        self.initial_data = getattr(serializer, 'initial_data', None)
+        self.instance = getattr(serializer, 'instance', None)
+
+    def __call__(self, value, *args, **kwargs):
+        status = value.get('status', self.instance.status)
+
+        if self.instance.status != status and status == 2:
+            self.validate_template_items(self.instance.templates.all()[0].template_items, value)
```
Comment (dmorina) [inline crowdsourcing/validators/project.py]: same as above

File: crowdsourcing/validators/project.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        for item in items.all():
+            if item.role == "input" or item.type == 'iframe':
+                has_input_item = True
+                break
+
+        if not has_input_item:
+            raise ValidationError('At least one input template item is required')
+
+        if self.instance.batch_files.count() > 0 and self.has_csv_linkage(
+                self.instance.templates.all()[0].template_items):
```
Comment (dmorina) [inline crowdsourcing/validators/project.py]: another one

File: crowdsourcing/viewsets/project.py [hunk 5]
```diff
@@ -258,3 +258,33 @@ def delete_file(self, request, **kwargs):
         else:
             models.ProjectBatchFile.objects.filter(batch_file_id=batch_file, project_id=kwargs['pk']).delete()
         return Response(data={}, status=status.HTTP_204_NO_CONTENT)
+
+    @list_route(methods=['GET'])
+    def requester_projects(self, request, **kwargs):
```
Comment (dmorina) [inline crowdsourcing/viewsets/project.py]: exclude this file i guess, have you merged?

**Comment by dmorina** [comment]
screen shot 2016-06-09 at 11 00 46 am

self.instance.status will fail on `create` because the instance does not exist yet

**Pull request merged**